### PR TITLE
[fix/viewer] Viewer fixes, refactoring and minor improvements

### PR DESCRIPTION
**Pull request by felix-schwarz**
## Description
- fixes various glitches and bugs when displaying files.
- refactoring of `DisplayViewController` and `DisplayHostViewController` for structural clarity, reinforcement and separation of functionality and concerns
- also fixes a few unrelated warnings
- asks the user before updating PDF files

## Related Issue
#935 #942  

## Screenshots (if appropriate):
Bildschirmfoto 2021-04-20 um 17 29 37

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## QA

checks: https://github.com/owncloud/ios-app/pull/951#issuecomment-826797194

reports:

- [ ] (1) Loading status: https://github.com/owncloud/ios-app/pull/951#issuecomment-826800324

### CODE REVIEW

**Comment by hosy** [comment]
@felix-schwarz unfortunately I ran into the same problem:
Attached the Xcode logs:
[file-could-not-be-opened-log.txt](https://github.com/owncloud/ios-app/files/6349750/file-could-not-be-opened-log.txt)

Simulator Screen Shot - iPad Air (4th generation) - 2021-04-21 at 11 33 58

**Comment by hosy** [comment]
@felix-schwarz what is strange: After running in the problem that the file could not be opened, I did the following:

- Tap `Manage` on the affected Account
- `Delete Local Copies`
- Opened the Account
- Opened the file `ownCloud Manual.pdf` and I will get the same error message

I expected that deleting all local copies will download the file again and could be opened.

**Comment by hosy** [comment]
@felix-schwarz I also discovered a crash:

Bildschirmfoto 2021-04-21 um 13 49 04

**Comment by felix-schwarz** [comment]
@hosy Thanks for the log. I'll go through it right away.

Regarding the crash, I found no smoking gun in the code at this place (especially the highlighted `self.class`) or in the lines that call it. Did you compile with `Debug` or `Release`?

**Comment by felix-schwarz** [comment]
@hosy I traced the path of the downloaded file and it doesn't seem to be removed. Yet it still mysteriously seems to vanish - or we're running into some weird filesystem encoding / normalization issues here:

```
2021-04-21 11:32:22.762000+0200 ownCloud[6507.13439886] 🛑 | [APP] Error fetching last modification date of file:///Users/matthias/Library/Developer/CoreSimulator/Devices/363E3717-921E-4F48-93F1-FA7510A604A3/data/Containers/Shared/AppGroup/4CB538B3-8690-420E-9CD2-86CF822FEE4C/File(null)rovider(null)torage/20FDA67C-3ECF-4A51-BBAC-34FC002FE32F/8C926AC7796B46BD8B90E0703C95234B/ownCloud2anual.pdf: Error Domain=NSCocoaErrorDomain Code=260 "The file “ownCloud Manual.pdf” couldn’t be opened because there is no such file." UserInfo={NSURL=file:///Users/matthias/Library/Developer/CoreSimulator/Devices/363E3717-921E-4F48-93F1-FA7510A604A3/data/Containers/Shared/AppGroup/4CB538B3-8690-420E-9CD2-86CF822FEE4C/File(null)rovider(null)torage/20FDA67C-3ECF-4A51-BBAC-34FC002FE32F/8C926AC7796B46BD8B90E0703C95234B/ownCloud2anual.pdf, NSFilePath=/Users/matthias/Library/Developer/CoreSimulator/Devices/363E3717-921E-4F48-93F1-FA7510A604A3/data/Containers/Shared/AppGroup/4CB538B3-8690-420E-9CD2-86CF822FEE4C/File Provider Storage/20FDA67C-3ECF-4A51-BBAC-34FC002FE32F/8C926AC7796B46BD8B90E0703C95234B/ownCloud Manual.pdf, NSUnderlyingError=0x7ffc352a7a30 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}} [DisplayViewController.swift:559|FULL]
```

I noticed these defects in the logged URL: **(null)**rovider**(null)**torage/20FDA67C-3ECF-4A51-BBAC-34FC002FE32F/8C926AC7796B46BD8B90E0703C95234B/ownCloud**2**anual.pdf

Can you also reproduce on a real device?

**Comment by hosy** [comment]
> @hosy Thanks for the log. I'll go through it right away.
> 
> Regarding the crash, I found no smoking gun in the code at this place (especially the highlighted `self.class`) or in the lines that call it. Did you compile with `Debug` or `Release`?

`Debug`

**Comment by felix-schwarz** [comment]
@hosy Thanks again for the log. It helped track this issue down, which turned out to be a really very specific edge case. I wrote down what happened in a new comment: https://github.com/owncloud/ios-sdk/blob/7abe23141d7a183add451e8770fc38592996f0c5/ownCloudSDK/Core/ItemList/OCCore%2BItemList.m#L570

The issue should finally be fixed now.

File: ownCloud/Client/Viewer/DisplayViewController.swift [hunk 1]
```diff
@@ -364,15 +345,57 @@ class DisplayViewController: UIViewController, Themeable {
...
+		case neverUpdate
+	}
+	var updateStrategy : UpdateStrategy = .alwaysUpdate
+
+	func shouldRenderItem(item: OCItem, isUpdate: Bool, shouldRender: @escaping (Bool) -> Void) {
+		if isUpdate {
+			switch updateStrategy {
+				case .ask:
+					OnMainThread {
+						let alert = UIAlertController(title: NSString(format: "%@ was updated".localized as NSString, item.name ?? "File".localized) as String, message: "Would you like to view the updated version?".localized, preferredStyle: .alert)
```
Comment (hosy) [inline ownCloud/Client/Viewer/DisplayViewController.swift:361]: `File" is missing in localized strings table

File: ownCloud/Client/Viewer/README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+- instantiates and provides the `DisplayViewController`s for the items in the page view controller
+- keeps track of changes in the active `DisplayViewController` to coordinate UI updates, especially in the navigation bar, and keep them consistent with dynamic changes from users paging through items
+- observes a new or existing`OCQuery` (typically from a file list view controller) for changes to automatically determine and update the list of items to present
+- handling of special cases like files being deleted from the server while being viewed, maintaining a natural pagination order regardless
+
+### `DisplayExtension`
+
+`DisplayExtension`s plug into the `OCExtension` system and are used to:
+- find the best matching viewer for a file
+- create an instance of the respective  `DisplayViewController` subclass
```
Comment (hosy) [inline ownCloud/Client/Viewer/README.md:38]: Awesome description! Thanks a lot!

**Review by hosy — CHANGES_REQUESTED** [review]

File: ownCloud/Client/Viewer/README.md (hunk 2 above)
Comment (felix-schwarz) [reply ownCloud/Client/Viewer/README.md:38]: Thanks!

**Comment by hosy** [comment]
@felix-schwarz thanks a lot! Yes, it fixes the issue.
BTW: The refactored code looks great!

Just found a very rare problem:
- add the following csv item to dem.owncloud.com:
[username.csv](https://github.com/owncloud/ios-app/files/6358820/username.csv)
- open the item on your device in preview
- open the item inside the browser in edit mode
- begin to add rows or change values

Mostly the item will be reloaded on the device, but sometimes the preview controller will stuck in "Loading…" state. 
Not sure if this is a bug/crash inside the `QLPreviewController`.

File: ownCloud/Client/Viewer/DisplayViewController.swift (hunk 1 above)
Comment (felix-schwarz) [reply ownCloud/Client/Viewer/DisplayViewController.swift:361]: Good catch! Added that. Thanks.

**Review by hosy — APPROVED** [review]

**Comment by felix-schwarz** [comment]
@hosy Could reproduce the CSV issue after many tries and found a deadlock that could cause a hang of the main thread + OCCore thread. Fixed the deadlock now. Hope it was the issue you ran into, too.

**Comment by jesmrec** [comment]
### QA

New PDF updating feature

- [X] Show new version
- [X] Refresh without asking
- [X] Repeat "Refresh without asking" to assure it woks with more modifications
- [X] Ignore updates

Other updates:

- [X] Update txt
- [X] Update image

**Comment by jesmrec** [comment]
### (1)

Probably is the same issue described above with CSV, but i have reproduced it with a txt file. I did the following steps:

1. Open a txt file in the device
2. Open the same file in the browser
3. Modify the txt file in the browser several times, and check the changes in the app

Finally, the app runs into a `Loading...` status

I caught some logs: [ownCloud_26_Apr_2021_at_14_26_00.log.txt](https://github.com/owncloud/ios-app/files/6377023/ownCloud_26_Apr_2021_at_14_26_00.log.txt)

app: `939956a`
sdk: `4483fd5`
iPhoneXR v14.4

**Comment by felix-schwarz** [comment]
@jesmrec I have been able to reproduce this on the device, but not in the Simulator and implemented a partial fix that creates a new `QLPreviewController` on every change rather than trying to reload the item.

Partial fix because the issue can still occur, but seems to be internal to `QLPreviewController` as the file exists, the file contents is also correct, but `QLPreviewController` simply doesn't finish.

Apparently others have run into this issue before: https://stackoverflow.com/questions/13649555/invalidating-qlpreviewcontroller-cache

I also tried to create a clone of the file with every change and then point `QLPreviewController` to that file, so it gets new versions under new URLs, so caching is not an issue. However, that didn't succeed either, with `QLPreviewController`  only showing file name + file size information then (Test.txt here, with 4 bytes in it):

Screen Shot 2021-04-27 at 15 36 39

I therefore believe that we can't currently work around this issue in `QLPreviewController`, only decrease its likelihood.

**Comment by jesmrec** [comment]
ok, then the problem under the hood is not totally fixable. We should not forget this, just in case. As i commented, i only reproduce with txt files. As https://github.com/owncloud/ios-app/pull/942 is not reproducible on my side as i posted in QA post, we can move this forward and wait for feedback/logs in case someone else can reproduce it. Is this ok for you @hosy ?

**Comment by hosy** [comment]
@jesmrec yes, this works for me. I will also keep this issue in mind and we can move forward. Thanks!

**Review by jesmrec — APPROVED** [review]

**Pull request merged**